core/src/main/java/org/apache/iceberg/BaseIncrementalAppendScan.java

```diff
@@ -0,0 +1,177 @@
+    List<Snapshot> snapshots = appendsBetween(table(), fromSnapshotIdExclusive, toSnapshotIdInclusive);
+    if (snapshots.isEmpty()) {
+      return CloseableIterable.empty();
+    }
+
+    return planFilesFromSnapshots(snapshots);
+  }
+
+  @Override
+  public CloseableIterable<CombinedScanTask> planTasks() {
+    CloseableIterable<FileScanTask> fileScanTasks = planFiles();
+    CloseableIterable<FileScanTask> splitFiles = TableScanUtil.splitFiles(fileScanTasks, targetSplitSize());
+    return TableScanUtil.planTasks(splitFiles, targetSplitSize(), splitLookback(), splitOpenFileCost());
+  }
+
+  private CloseableIterable<FileScanTask> planFilesFromSnapshots(List<Snapshot> snapshots) {
+    Set<Long> snapshotIds = Sets.newHashSet(Iterables.transform(snapshots, Snapshot::snapshotId));
+    Set<ManifestFile> manifests = FluentIterable
+        .from(snapshots)
+        .transformAndConcat(Snapshot::dataManifests)
+        .filter(manifestFile -> snapshotIds.contains(manifestFile.snapshotId()))
+        .toSet();
+
+    ManifestGroup manifestGroup = new ManifestGroup(tableOps().io(), manifests)
+        .caseSensitive(context().caseSensitive())
+        .select(context().returnColumnStats() ? DataTableScan.SCAN_WITH_STATS_COLUMNS : DataTableScan.SCAN_COLUMNS)
+        .filterData(context().rowFilter())
+        .filterManifestEntries(
+            manifestEntry ->
+                snapshotIds.contains(manifestEntry.snapshotId()) &&
+                    manifestEntry.status() == ManifestEntry.Status.ADDED)
+        .specsById(tableOps().current().specsById())
+        .ignoreDeleted();
+
+    if (context().ignoreResiduals()) {
+      manifestGroup = manifestGroup.ignoreResiduals();
+    }
+
+    if (manifests.size() > 1 &&
+        (DataTableScan.PLAN_SCANS_WITH_WORKER_POOL || context().planWithCustomizedExecutor())) {
+      manifestGroup = manifestGroup.planWith(context().planExecutor());
+    }
+
+    return manifestGroup.planFiles();
+  }
+
+  private static void notifyIncrementalScanEvent(
+      Table table, TableScanContext context, long toSnapshotIdInclusive) {
+    long fromSnapshotId;
+    boolean fromSnapshotInclusive;
+    if (context.fromSnapshotId() != null) {
+      fromSnapshotId = context.fromSnapshotId();
+      fromSnapshotInclusive = context.fromSnapshotInclusive();
+    } else {
+      Snapshot oldestAncestorSnapshot = SnapshotUtil.oldestAncestor(toSnapshotIdInclusive, table::snapshot);
+      fromSnapshotId = oldestAncestorSnapshot.snapshotId();
+      fromSnapshotInclusive = true;
+    }
+
+    Listeners.notifyAll(new IncrementalScanEvent(table.name(), fromSnapshotId,
+        toSnapshotIdInclusive, context.rowFilter(), table.schema(), fromSnapshotInclusive));
+  }
+
+  private static List<Snapshot> appendsBetween(Table table, Long fromSnapshotIdExclusive, long toSnapshotIdInclusive) {
+    List<Snapshot> snapshots = Lists.newArrayList();
+    for (Snapshot snapshot : SnapshotUtil.ancestorsBetween(
```

Comment:
Comment: this does not validation because `planFiles` validated the from snapshot and to snapshot range.